test/e2e/lib/helpers/broker_event_transformation_test_with_source_helper.go

```diff
@@ -0,0 +1,331 @@
+
+func BrokerEventTransformationTestWithSchedulerSourceHelper(t *testing.T, client *lib.Client, authConfig lib.AuthConfig, brokerURL url.URL, brokerName string) {
+	data := "my test data"
+	targetName := "event-display"
+	sName := "scheduler-test"
+	// Create a target Job to receive the events.
+	lib.MakeSchedulerJobOrDie(client, data, targetName)
+	createTriggersAndKService(client, brokerName, targetName)
+
+	var url apis.URL = apis.URL(brokerURL)
+	// Just to make sure all resources are ready.
+	time.Sleep(5 * time.Second)
+
+	// Create the CloudSchedulerSource.
+	lib.MakeSchedulerOrDie(client, sName, data, targetName, authConfig.PubsubServiceAccount,
+		kngcptesting.WithCloudSchedulerSourceSinkURI(&url),
+	)
+
+	msg, err := client.WaitUntilJobDone(client.Namespace, targetName)
+	if err != nil {
+		t.Error(err)
+	}
+
+	t.Logf("Last termination message => %s", msg)
+	if msg != "" {
+		out := &lib.TargetOutput{}
+		if err := json.Unmarshal([]byte(msg), out); err != nil {
+			t.Error(err)
+		}
+		if !out.Success {
+			// Log the output of scheduler pods
+			if logs, err := client.LogsFor(client.Namespace, sName, lib.CloudSchedulerSourceTypeMeta); err != nil {
+				t.Error(err)
+			} else {
+				t.Logf("scheduler log: %+v", logs)
+			}
+
+			// Log the output of the target job pods
+			if logs, err := client.LogsFor(client.Namespace, targetName, lib.JobTypeMeta); err != nil {
+				t.Error(err)
+			} else {
+				t.Logf("addressable job: %+v", logs)
+			}
+			t.Fail()
+		}
+	}
+
+}
+
+func CreateKService(client *lib.Client) string {
+	kserviceName := helpers.AppendRandomString("kservice")
+	// Create the Knative Service.
+	kservice := resources.ReceiverKService(
+		kserviceName, client.Namespace)
+	client.CreateUnstructuredObjOrFail(kservice)
+	return kserviceName
+
+}
+
+func createTriggerWithKServiceSubscriber(client *lib.Client, brokerName, kserviceName string) {
+	dummyTriggerName := "dummy-broker-" + brokerName
+	client.Core.CreateTriggerOrFail(
+		dummyTriggerName,
+		eventingtestresources.WithBroker(brokerName),
+		eventingtestresources.WithAttributesTriggerFilter(
+			eventingv1alpha1.TriggerAnyFilter, eventingv1alpha1.TriggerAnyFilter,
+			map[string]interface{}{"type": "e2e-testing-dummy"}),
+		eventingtestresources.WithSubscriberServiceRefForTrigger(kserviceName),
+	)
+}
+
+func createTriggerWithTargetServiceSubscriber(client *lib.Client, brokerName, targetName string) {
+	respTriggerName := "resp-broker-" + brokerName
+	client.Core.CreateTriggerOrFail(
+		respTriggerName,
+		eventingtestresources.WithBroker(brokerName),
+		eventingtestresources.WithAttributesTriggerFilter(
+			eventingv1alpha1.TriggerAnyFilter, eventingv1alpha1.TriggerAnyFilter,
+			map[string]interface{}{"type": "e2e-testing-resp"}),
+		eventingtestresources.WithSubscriberServiceRefForTrigger(targetName),
+	)
+}
+
+func createTriggersAndKService(client *lib.Client, brokerName, targetName string) {
+	// Create the Knative Service.
+	kserviceName := CreateKService(client)
```

Comment:
Let's move this out for consistency - the targetName is created outside of this function. Let's make this function only create triggers, not target services

Reply:
This function is used for all GCP Broker with different sources. That's why we move it here.
We already have specific function for creating triggers.

Reply:
Then I don't think we need this function. I would just inline CreateKService and CreateTrigger

Reply:
If we don't have this function, all test cases will need to inline these 5 codes. That's why I change it to one function. For the above comments, I did't create anotherfunc createTrigger(client *lib.Client, brokerName, triggerName, extensions) is because this function will end up calling just one line 
`client.Core.CreateTriggerOrFail(
                 triggerName,
		eventingtestresources.WithBroker(brokerName),
		eventingtestresources.WithAttributesTriggerFilter(
			eventingv1alpha1.TriggerAnyFilter, eventingv1alpha1.TriggerAnyFilter,
			map[string]interface{}{"type": "e2e-testing-resp"}),
		eventingtestresources.WithSubscriberServiceRefForTrigger(targetName),
	)`

Reply:
My main point about this one is it's unclear to me what it does but reading its name. I feel better if the KService is created outside of this function, and this is changed to `createTriggersForKserviceAndTargetService`...

But I would'd say this is a blocker.

Reply:
I will change the naming  to `createTriggersForKserviceAndTargetService`